CIP-0118/README-nested.md

```diff
@@ -0,0 +1,538 @@
+we discuss here, we allow transactions that specify
+_a swap request_. Underspecified transaction bodies are included in the 
+body of a _top-level_ transactions (which may itself be partially specified) alongside
+additional required data (e.g. their redeemers). We call the 
+set of transactions containing the top-level transaction and the _sub-transactions_
+it includes a _batch_. Applying a complete batch results in a valid ledger update, 
+however, applying each of the individual transactions may not. For example, insufficient 
+fees or collateral may be provided, missing witness info, etc.
+In the Babel-fees usecase, transactions do not 
+need to be balanced, i.e they do not require specifying where any missing funds come from, 
+or any extra funds go. A complete batch, however, must be balanced.
+
+## Motivation: why is this CIP necessary?
+
+This CIP provides a partial solution to the problems described in 
+[CPS-15](https://github.com/cardano-foundation/CIPs/pull/779).
+In particular, it describes some ledger changes that allow intent settlement
+of intents that require “counterparty irrelevance”, including many of the swap use cases
+and dApp fee sponsorship. The motivation behind designing this solution is _Babel fees_,
+which is a requirement to support "paying transaction fees in a non-Ada currency".
+The Babel-fees usecase is a special case of a swap. Supporting swaps is very desirable functionality for cryptocurrency
+ledgers.
+
+There are off-chain solutions being developed for the purpose of achieving the same 
+(major) goals, including fee coverage, atomic swaps, DApp fee sponsorship, 
+collateral sponsorship, etc. This gives us indication that it is a worthy pursuit 
+to solve this problem at the ledger level, in a way that is cheaper, more convenient, 
+maintains formal guarantees, requires less off-chain communication, and does not require deposits or interaction 
+with smart contracts. 
+
+The ledger changes we describe have been developed as a result of the discussions and 
+proposals in previous CIPs, including previous versions of validation zones proposals, 
+and Transaction Swaps. The following 
+new usecases and functionality (in addition to batching) are supported by the changes :
+
+1. transactions in a single batch may pay each others' fees and `minUTxOValue` ;
+2. sharing (for the purpose of deduplication) of scripts across transactions 
+in a single batch ;
+3. scripts may inspect the contents of other transactions within the same batch ;
+4. the top-level transaction may provide collateral scripts in its sub-transactions ;
+5. a new intent we call "spend-by-output", wherein a sub-transactions may specify _outputs_ it 
+intends to spend, and the top-level transaction specifies the inputs that point 
+to those outputs in the UTxO set. This is included as a way to showcase that this 
+change to the ledger rules establishes the infrastructure to add additional 
+kids of supported intents (see CPS-15).
```

Comment:
I don't understand what this means. The use of the terms "inputs" and "outputs" is confusing to me. In my vocabulary, "inputs" are UTxOs that a transaction intends to spend while "outputs" are UTxOs that would be created by the transaction. So I don't understand what it means to "specify outputs it intends to spend" or "[specify] the inputs that point to those outputs in the UTxO set". Can you explain what you mean by this?

Reply:
My interpretation is that "output" is the contents (address/datum/value), while "input" is the out_ref pointing to a specific output.

> 5. a new intent we call "spend-by-output", wherein a sub-transaction may specify the contents of the outputs it intends to spend, and the top-level transaction specifies the transaction references to those output in the UTxO set

Reply:
suppose `tx` has inputs `txin1, txin2` and output `o1`

also, in the UTxO we have 

`txin1 |-> o`\
`txin2 |-> o'`\

then, we can build `tx'` that has `spendOuts` containing `o, o'` and output `o1`. A top-level transaction builder can then complete the batch with `txTop`, with (possibly some inputs), and 

`corInputs` containing `txin1, txin2`

Then, in some sense, `tx` and `tx'; txTop` do a similar thing - remove entries corresponding to `txin1, txin2` from the UTxO, and add `o, o'` to the UTxO. `tx'` would likely include a payment to whoever builds `txTop` for their tx-building services.

The reason we are proposing this is that if a light client (LC) is not following the chain, they would be unaware of the `txin`s they would want to spend. If an LC requests a service provider to build a transaction for them that satisfies some specification (e.g. "pay key `k` the amount `x` of ada from key `k'`), the SP should be able to respond in a way that the LC does not use the data in this response to construct a new valid transaction that excludes payment to the SP. Obscuring inputs that SP can supply later makes it so that the transaction is not valid unless it comes with a top-level tx that provides the missing inputs.